dag_in_context/src/optimizations/memory.egg

```diff
@@ -0,0 +1,402 @@
+      :ruleset always-run)
+(rule ((PointsToCells (Alloc id sz state ty) ap))
+      ((PointsToCells state ap))
+      :ruleset always-run)
+
+; Compute and propagate PointsToCells
+(rewrite (PointsToCells (Concat x y) aps)
+         (TuplePointsTo (Concat-List<PtrPointees>
+           (UnwrapTuplePointsTo (PointsToCells x aps))
+           (UnwrapTuplePointsTo (PointsToCells y aps))))
+         :ruleset always-run)
+
+(rewrite (PointsToCells (Get x i) aps)
+         (GetPointees (PointsToCells x aps) i)
+         :ruleset always-run)
+
+(rewrite (PointsToCells (Single x) aps)
+         (TuplePointsTo
+           (Cons-List<PtrPointees>
+             (UnwrapPtrPointsTo (PointsToCells x aps))
+             (Nil-List<PtrPointees>)))
+         :ruleset always-run) 
+
+(rewrite (PointsToCells (Arg ty) aps)
+         aps
+         :ruleset always-run)
+
+; Allow non-pointer types to resolve
+(rule ((PointsToCells x aps)
+       (HasType x ty))
+      ((TypeToPointees ty))
+      :ruleset always-run)
+(rule ((= f (PointsToCells x aps))
+       (HasType x ty)
+       (= pointees (TypeToPointees ty))
+       (PointsNowhere pointees))
+      ((union f pointees))
+      :ruleset always-run)
+
+(rewrite (PointsToCells (Bop (PtrAdd) x e) aps)
+         (PtrPointsTo
+           (AddIntIntervalToPtrPointees
+             (MkIntInterval (I lo) (I hi))
+             (UnwrapPtrPointsTo (PointsToCells x aps))))
+         :when ((= (IntB lo) (lo-bound e))
+                (= (IntB hi) (hi-bound e)))
+         :ruleset always-run)
+
+(rewrite (PointsToCells (If inputs c t e) aps)
+         (UnionPointees
+           (PointsToCells t (PointsToCells inputs aps))
+           (PointsToCells e (PointsToCells inputs aps)))
+         :ruleset always-run)
+
+(rewrite (PointsToCells (Alloc id sz state ty) aps)
+         (TuplePointsTo
+           (Cons-List<PtrPointees>
+             (PointsTo
+               (Cons-List<i64+IntInterval>
+                 id
+                 (MkIntInterval (I 0) (I 0))
+                 (Nil-List<i64+IntInterval>)))
+             (Cons-List<PtrPointees>
+               (PointsTo (Nil-List<i64+IntInterval>)) ; state output points to nothing
+               (Nil-List<PtrPointees>))))
+         :ruleset always-run)
+
+; arg pointees * loop in * loop out * i64 -> result pointees
+(function PointsToCellsAtIter (Pointees Expr Expr i64) Pointees)
+
+; compute first two
+(rule ((PointsToCells (DoWhile inputs pred-body) aps))
+      ((set (PointsToCellsAtIter aps inputs pred-body 0)
+            (PointsToCells inputs aps))
+       (set (PointsToCellsAtIter aps inputs pred-body 1)
+            (UnionPointees
+              (PointsToCellsAtIter aps inputs pred-body 0)
+              (PointeesDropFirst
+                (PointsToCells pred-body (PointsToCellsAtIter aps inputs pred-body 0))))))
+      :ruleset always-run)
+
+; Note that this rule is bounded by ruleset memory
+(rule ((= pointees0 (PointsToCellsAtIter aps inputs pred-body i))
```

Comment:
This might be quadratic in # of iters since egglog is not smart enough to use `i` to look up the value of `(+ i 1)`. Instead it goes through the cross product (i, j) and check if `i + 1 == j`... Maybe we can materialize a `succ` relation...

Reply:
Added!